File: Tree/BST/bst.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int data) : data(data), left(nullptr), right(nullptr) {}
};

Node* insert(Node* root, int val) {
    if (root == nullptr) {
        root = new Node(val);
        return root;
    }
    if (val < root->data) {
        root->left = insert(root->left, val);
    } else {
        root->right = insert(root->right, val);
    }
    return root;
}

bool search(Node* root, int key) {
    if (root == nullptr) return false;
    if (key == root->data) return true;
    if (key < root->data) {
        return search(root->left, key);
    }
    return search(root->right, key);
}

Node* deleteNode(Node* root, int key) {
    if (!search(root, key)) return root;
    if (key < root->data) {
        root->left = deleteNode(root->left, key);
    } else if (key > root->data) {
        root->right = deleteNode(root->right, key);
    } else {
        if (root->left == nullptr && root->right == nullptr) {
            delete root;
            return nullptr;
        }
        if (root->left == nullptr) {
            Node* temp = root->right;
            delete root;
            return temp;
        }
        if (root->right == nullptr) {
            Node* temp = root->left;
            delete root;
            return temp;
        }

        Node* inOrderSuccessor = findInOrderSuccessor(root->right);
        root->data = inOrderSuccessor->data;
        root->right = deleteNode(root->right, inOrderSuccessor->data);
    }
    return root;
}

Node* findInOrderSuccessor(Node* root) {
    while (root->left != nullptr) {
        root = root->left;
    }
    return root;
}

void inorder(Node* root) {
    if (root == nullptr) return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}

int main() {
    int values[] = {5, 1, 3, 4, 2, 7};
    Node* root = nullptr;

    for (int i : values) {
        root = insert(root, i);
    }
    
    inorder(root);
    cout << endl;

    root = deleteNode(root, 0);
    inorder(root);

    return 0;
}

File: Tree/BST/bst.java
public class bst{
    static class Node{
        int data;
        Node left;
        Node right;

        Node(int data){
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }
    public static Node insert(Node root,int val){
        if(root==null){
            root = new Node(val);
            return root;
        }
        if(root.data>val){
            root.left = insert(root.left, val);
        }
        else{
            root.right = insert(root.right, val);
        }
        return root;
    }
    public static boolean search(Node root,int key){
        if(root==null) return false;
        if(key==root.data) return true;
        if(key<root.data){
            return search(root.left, key);
        }
        return search(root.right, key);
    }
    public static Node createBST(int arr[],int start,int end){
        if(start>end) return null;
        int mid = (start+end)/2;
        Node root = new Node(arr[mid]);
        root.left = createBST(arr, start, mid-1);
        root.right = createBST(arr, mid+1, end);
        return root;
    }
    public static Node delete(Node root, int key){
        if(!search(root, key)) return root;
        if(key<root.data){
            root.left =  delete(root.left, key);
        }
        else if(key>root.data){
            root.right = delete(root.right, key);
        }
        else{
            if(root.left==null && root.right==null){
                return null;
            }
            if(root.left==null) return root.right;
            if(root.right==null) return root.left;

            Node inOrderSuccessor = findInOrderSuccessor(root.right);
            root.data = inOrderSuccessor.data;
            root.right = delete(root.right, inOrderSuccessor.data);
        }
        return root;
    }
    public static Node findInOrderSuccessor(Node root){
        while(root.left!=null){
            root = root.left;
        }
        return root;
    }
    public static void inorder(Node root){
        if(root==null) return;
        inorder(root.left);
        System.out.print(root.data+" ");
        inorder(root.right);
    }
    
    public static void main(String[] args) {
        // int values[] = {5,1,3,4,2,7};
        // Node root = null;

        // for(int i:values){
        //     root = insert(root, i);
        // }
        // inorder(root);
        // System.out.println();
        // inorder(delete(root, 0));
        int arr[] = {3,5,6,8,10,11,12};
        Node root = createBST(arr, 0, arr.length-1);
        inorder(root);
    }
}